docs/design/api-guidelines/api-guidelines.md

```diff
@@ -0,0 +1,233 @@
+- [RFC 7231](https://tools.ietf.org/html/rfc7231)
+
+# Conventions & Rules
+
+### Resource
+
+URIs follow [RFC 3986](https://tools.ietf.org/html/rfc3986) specification.
+
+#### URI Naming Conventions
+
+[Resource Naming](https://www.restapitutorial.com/lessons/restfulresourcenaming.html)
+
+* Use nouns and NOT the verbs.
+* Use the plural version of a resource name to be consistent when referring to particular resources, e.g.:
+```
+/switches
+/flows
+/links
+```
+* Use lowercase and dash-separated path names, e.g.:
+```
+/health-check
+```
+* Use snake_case (underscore-separated) query string literals.
+
+### Use HTTP methods to operate on collections and entities
+
+[HTTP methods](https://www.restapitutorial.com/lessons/httpmethods.html)
+
+E.g.
+
+| Resource / HTTP method | POST (create)    | GET (read)  | PUT (update)           | DELETE (delete)    |
+| ---------------------- | ---------------- | ----------- | ---------------------- | ------------------ |
+| /flows                 | Create new flow  | List flows  | Error                  | Error              |
+| /flows/{flow_id}       | Error            | Get flow    | Update flow if exists  | Delete flow        |
+
+
+Update and create must return a resource representation.
+
+#### Create entity
+
+* Server returns the `HTTP 201 Created` status code
+* Response will also include the `Location` header that points to the URL of the new resource ([RFC 6570](http://tools.ietf.org/html/rfc6570))
+* Server returns the new entity
+
+#### List entities (collection)
+
+* Response collection is presented as a plain array, not a dictionary with the array as a value.
+
+#### Update entity
+
+* To update an entity, the server requires being provided with all the properties that can be subject to update
+* Server returns an updated entity
+
+#### Delete entity
+
+* Server returns the `HTTP 204 No Content` status code if response content is empty.
+
+### Return appropriate status codes
+
+[HTTP Status Codes](http://www.restapitutorial.com/httpstatuscodes.html)
+
+Return appropriate HTTP status codes with each response.
+Responses concerning successful operations should be coded according to the following rules:
+
+* `200 Ok`: Successful execution of `GET`, `PUT`, `PATCH` request. An existing resource is updated synchronously.
+* `201 Created`: Response to a `POST` request that results in entity creation.
+* `202 Accepted`: Accepted request concerning a `POST`, `PUT`, `DELETE`, or `PATCH` request that will be processed asynchronously.
+* `204 No Content`: Response to a successful request that will return empty body (as a `DELETE` request)
+
+Pay attention to the use of authentication and authorization error codes:
+
+* `401 Unauthorized`: Request failed because user is not authenticated
```

Comment:
I would add the article 'a' before 'user'.

Reply:
Fixed.